src/transformStateWithClones.js

```diff
@@ -7,7 +7,32 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const transformStateResult = [];
+  const stateClone = { ...state };
+
+  for (const action of actions) {
+    if (action.hasOwnProperty('type') && action.type === 'clear') {
+      const keys = Object.keys(stateClone);
+
+      for (const key of keys) {
+        delete stateClone[key];
+      }
+    }
+
+    if (action.hasOwnProperty('type') && action.type === 'addProperties') {
+      Object.assign(stateClone, action.extraData);
+    }
+
+    if (action.hasOwnProperty('type') && action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete stateClone[key];
+      }
+    }
+
+    transformStateResult.push({ ...stateClone });
```

Comment:
Pushing a copy of the `stateClone` object to the `transformStateResult` array should be done after processing each action. Ensure that the state is cloned at the beginning of each loop iteration to maintain the integrity of the state for each action.